GitUp/Application/WelcomeWindow.m

```diff
@@ -97,33 +97,125 @@ - (void)concludeDragOperation:(id<NSDraggingInfo>)sender {
+@property (weak, nonatomic, readonly) WelcomeWindow *welcomeWindow;
+@property (strong, nonatomic, readwrite) WelcomeWindowControllerModel *model;
 @end
 
-@implementation WelcomeWindow
+@implementation WelcomeWindowController
+- (WelcomeWindow *)welcomeWindow {
+  return (WelcomeWindow *)self.window;
+}
+#pragma mark - Initialization
+- (instancetype)init {
+  return [[super initWithWindowNibName:@"Welcome"] configuredWithModel:[[WelcomeWindowControllerModel alloc] init]];
+}
 
-#pragma mark - Accessors
-- (AppDelegate *)appDelegate {
-  return [AppDelegate sharedDelegate];
+#pragma mark - Setup
+- (void)setupUIElements {
+  self.twitterButton.textAlignment = NSLeftTextAlignment;
+  self.twitterButton.textFont = [NSFont boldSystemFontOfSize:11];
+  self.forumsButton.textAlignment = NSLeftTextAlignment;
+  self.forumsButton.textFont = [NSFont boldSystemFontOfSize:11];
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(willShowPopUpMenu) name:NSPopUpButtonWillPopUpNotification object:self.recentPopUpButton];
+  
+  self.closeButton.action = @selector(closeButtonPressed);
+  self.twitterButton.action = @selector(openTwitter);
+  self.forumsButton.action = @selector(viewIssues);
+  
+  self.closeButton.target = self;
+  self.twitterButton.target = self;
+  self.forumsButton.target = self;
+}
+
+#pragma mark - Window Lifecycle
+- (void)windowDidLoad {
+  [super windowDidLoad];
+  WelcomeWindow *welcomeWindow = (WelcomeWindow *)self.window;
+  welcomeWindow.configureItem = self.model.configureItem;
+  welcomeWindow.getRecentDocuments = self.model.getRecentDocuments;
+  
+  [self setupUIElements];
+}
+
+#pragma mark - Configuration
+- (instancetype)configuredWithModel:(WelcomeWindowControllerModel *)model {
+  self.model = model;
+  return self;
+}
+
+#pragma mark - Reactions
+- (void)handleDocumentCountChanged {
+  BOOL showWelcomeWindow = self.model.getUserDefaultsShouldShow();
+  if (showWelcomeWindow && (self.model.shouldShow) && ![[[NSDocumentController sharedDocumentController] documents] count]) {
+    if (!self.welcomeWindow.visible) {
+      [self.welcomeWindow makeKeyAndOrderFront:nil];
+    }
+  } else {
+    if (self.welcomeWindow.visible) {
+      [self.welcomeWindow orderOut:nil];
+    }
+  }
 }
 
 #pragma mark - Actions
+- (void)closeButtonPressed {
+  [self.model setShouldHide];
+  [self.window orderOut:nil];
+}
+
+#pragma mark - Actions/Recent
 - (void)cleanupRecentEntries {
   NSMenu* menu = self.recentPopUpButton.menu;
   while (menu.numberOfItems > 1) {
     [menu removeItemAtIndex:1];
   }
 }
+
 - (void)willShowPopUpMenu {
   [self cleanupRecentEntries];
   NSMenu* menu = self.recentPopUpButton.menu;
-  NSArray* array = self.getRecentDocuments(); // [[NSDocumentController sharedDocumentController] recentDocumentURLs];
+  NSArray* array = self.model.getRecentDocuments(); // [[NSDocumentController sharedDocumentController] recentDocumentURLs];
```

Comment:
Should really just call the commented code `[[NSDocumentController sharedDocumentController] recentDocumentURLs];`. Makes it much clearer.

Reply:
@lucasderraugh Fixed.